### ISSUE TITLE
display gene table and heatmap

### ISSUE THREAD

**Issue by slowkow** [maximilianh/cellBrowser#26]
Right now the table of genes with p-values is only visible after clicking on the name of a cluster.

This means you have to choose which display you want to see, either the map of cells or the table of genes, but not both at the same time.

You might consider an alternative display, as in the [Loupe] software by 10X Genomics.

[Loupe]: https://support.10xgenomics.com/single-cell-gene-expression/software/visualization/latest/what-is-loupe-cell-browser

Here are two screenshots that show how they do it:

![image](https://user-images.githubusercontent.com/209714/45762635-cd944880-bbfc-11e8-8189-d253fd9afb1e.png)

![image](https://user-images.githubusercontent.com/209714/45762694-ef8dcb00-bbfc-11e8-86d2-0cfdd72cb5c5.png)

I like the interactive table:

- When you click on a cluster name in the table header, you get to see the genes ranked by p-value in that cluster.
- The other clusters that are out-of-focus only show a summarized Log2 Fold-Change, not the p-value.

What do you think?

**Comment by maximilianh** [comment]
This is an excellent idea. I use Loupe but for some reason never had the
idea of copying this layout. I can't promise that I'll do it right away,
but the next time when I'm playing with the UI, I'll certainly implement
this. Thanks!

**Comment by maximilianh** [comment]
still didn't get to do this. other users have requested violin plots of selected marker genes on the clusters rather than a heatmap. Opinion?

**Comment by slowkow** [comment]
A violin is better than nothing, but I have opinions after seeing many figures.

Violins are not appropriate for sparse data -- most cells have 0 expression for most genes. Violins show a weird blob at the bottom, creating a false impression of low levels of expression.

For this reason, I like to show a bar that shows how many cells are nonzero, giving the additional information absent from the violin.

I use `geom_quasirandom()` from the [ggbeeswarm] package to show only the non-zero values, ignoring all of the zeros that are shown by the bar.

[ggbeeswarm]: https://github.com/eclarke/ggbeeswarm

<img src="https://user-images.githubusercontent.com/209714/53652403-cf841e00-3c16-11e9-8308-e03fd08ae04c.png" width="40%">

A heatmap where each row is a gene and each column is a cell is a valuable way to see subclusters that are not represented by a 2D map (tSNE, UMAP, PCA, etc.). It's also valuable to see correlated genes to your query gene.

**Comment by maximilianh** [comment]
thanks, very nice idea to remove the 0s!!

oopps... every column is a cell? that won't be easy, most of the more
recent 10x datasets have >30k cells now...

I was thinking of doing a heatmap cluster x genes...

**Comment by slowkow** [comment]
You're right that showing cells only works for small data. A heatmap of clusters (columns) and genes (rows) would be great, too.
